### PR TITLE
File-like objects support for h5py.File

### PR DESCRIPTION
**Pull request by aparamon**
Please review the initial commit.
Help needed: how to access H5FD_t/H5FD_class_t structs from Python? How to create H5FD_class_t struct from Python and populate it with Python functions?

### CODE REVIEW

**Comment by aparamon** [comment]
Initial implementation of https://github.com/h5py/h5py/issues/1006 is complete!
Please review; hopefully the interface and tests are good enough. One thing to be improved is better error handling inside Cython handlers -- I'll add that soon.

File: h5py/_hl/files.py [hunk 1]
```diff
@@ -279,17 +279,18 @@ def swmr_mode(self, value):
             else:
                 raise ValueError("It is not possible to forcibly switch SWMR mode off.")
 
-    def __init__(self, name, mode=None, driver=None,
+    def __init__(self, file, mode=None, driver=None,
```
Comment (aragilar) [inline h5py/_hl/files.py]: I would avoid changing `name` to `file`, as there may be users using `name` via keyword arguments.

**Comment by aragilar** [comment]
Apart from the name change, this looks reasonable enough to me (I haven't looked in detail at the low-level code, which may be the source of the test failures). It appears that tests are failing on appveyor (windows) on py27, so that'll need fixing also.

File: h5py/h5fd.pyx [hunk 2]
```diff
@@ -77,3 +77,79 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+
+cdef herr_t H5FD_fileobj_flush(H5FD_fileobj_t *f, hid_t dxpl, hbool_t closing):
+    (<object>f.fileobj).flush()
+    return 0
+
+
+cdef H5FD_class_t info
+memset(&info, 0, sizeof(info))
+
+info.name = 'fileobj'
```
Comment (aragilar) [inline h5py/h5fd.pyx]: I'm confused what's going on in this section, can we add some docs outlining what's going on here?

File: h5py/_hl/files.py [hunk 3]
```diff
@@ -411,3 +416,7 @@ def __repr__(self):
         if six.PY2:
             return r.encode('utf8')
         return r
+
+
+register_driver('fileobj',
+                lambda plist: plist.set_driver(h5fd.fileobj_driver))
```
Comment (aragilar) [inline h5py/_hl/files.py]: Shouldn't this been in h5py/h5fd.pyx?

File: h5py/_hl/files.py (hunk 1 above)
Comment (aparamon) [reply h5py/_hl/files.py]: Indeed, that might be undesirable.
[pyfive](https://github.com/jjhelmus/pyfive) uses "filename : str or file-like":
https://github.com/jjhelmus/pyfive/blob/46dc0bd19b9126d33e6b4eb7153451f3820103d3/pyfive/high_level.py#L152
Shall I just let `name` be a file-like object, or additionally introduce argument `file`, for cleanness?

File: h5py/h5fd.pyx (hunk 2 above)
Comment (aparamon) [reply h5py/h5fd.pyx]: Sure, I'll do. In short, I'm allocating a VFL driver descriptor that dispatches HDF5 I/O to `H5FD_fileobj*` procedures (which, in turn, translate that to Python file-like object method calls).

File: h5py/_hl/files.py (hunk 3 above)
Comment (aparamon) [reply h5py/_hl/files.py]: I'm not sure -- `register_driver()` was already defined in https://github.com/h5py/h5py/blob/master/h5py/_hl/files.py and seems perfectly suitable.

**Comment by aparamon** [comment]
I have noticed that "minideps" target specifies Cython version 0.23 explicitly. Is it possible that this is lifted, for example to Cython 0.28? Or, please suggest an older Cython idiom for implementation of `H5FD_fileobj_read`/`H5FD_fileobj_write` (basically, data transfer between `bytearray` and `void*`)

**Comment by aragilar** [comment]
`mindeps` is a check that the code still runs on minimum (oldest) numpy/cython versions given in the setup.py file. cython 0.28 is quite new (from this year), so hopefully we can avoid needing such a new cython. Do we need the `const` keywords, as that seems to be the reason for the test failures?

**Comment by aparamon** [comment]
Keyword `const` defines read-only view. Simply removing `const` leads to
````pytb
  File "stringsource", line 653, in View.MemoryView.memoryview_cwrapper
  File "stringsource", line 348, in View.MemoryView.memoryview.__cinit__
BufferError: Object is not writable.
````

**Comment by aparamon** [comment]
Hmmm, it's unclear to me why these are failing:
https://travis-ci.org/h5py/h5py/jobs/399555868
https://ci.appveyor.com/project/h5py/h5py/build/1.0.348/job/macd9n1o4j6wjqyq
https://ci.appveyor.com/project/h5py/h5py/build/1.0.348/job/m2o8wpn7b6ptpoee
Any clue?

**Comment by tacaswell** [comment]
The failures on windows map to (see https://blogs.msdn.microsoft.com/calvin_hsia/2004/06/30/what-is-a-c0000005-crash/ ) which suggests there is a memory bug someplace in the cython code...

**Comment by aparamon** [comment]
@tacaswell It makes sense. Is there a way I run that particular AppVeyor test locally, in order to bisect and fix the problem? It would be a poor idea to make the bisecting commits into this PR and wait for the AppVeyor to proceed...

**Comment by tacaswell** [comment]
If you have a windows machine set up for development you should be able to run the commands we run on appveyor locally.

**Comment by aparamon** [comment]
Yes I do! I have a lot of MS C compilers installed, so hopefully it serves well. I'm a bit hesitant to installing more Pythons, but since it's reproducible on 3.6 my default version should be Ok.
Where can I get the script?

**Comment by tacaswell** [comment]
https://github.com/h5py/h5py/blob/master/appveyor.yml is the input to appveyor and the support scripts are at https://github.com/h5py/h5py/tree/master/ci

**Comment by aparamon** [comment]
Hmm, the problem was with the format string. I tried 2 fixed, both not completely successful:

1.
````cython
    cdef ptruint_t int p
    if sscanf(name, "%zx", &p):
````
Works except for older (<14) MSVC complilers,

2.
````cython
    cdef unsigned long long int p
    if sscanf(name, "%llx", &p):
````
Works except for 32-bit.

Please suggest the best cross-platform way to convert hex to `Py_Object *`.

File: h5py/_hl/files.py [hunk 4]
```diff
@@ -342,15 +343,20 @@ def __init__(self, name, mode=None, driver=None,
             with phil:
                 fid = h5i.get_file_id(name)
         else:
-            name = filename_encode(name)
+            if hasattr(name, 'read') and hasattr(name, 'seek'):
+                name = hex(id(name)).encode('ASCII')
+                driver = 'fileobj'
```
Comment (tacaswell) [inline h5py/_hl/files.py]: Can you add a check there that the user did not pass in a non-default value for `driver` and raise if they did?

File: h5py/h5fd.pyx [hunk 5]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+cdef haddr_t H5FD_fileobj_get_eof(const H5FD_fileobj_t *f, H5FD_mem_t type) except -1:
+    (<object>f.fileobj).seek(0, SEEK_END)
+    return (<object>f.fileobj).tell()
+
+cdef herr_t H5FD_fileobj_read(H5FD_fileobj_t *f, H5FD_mem_t type, hid_t dxpl, haddr_t addr, size_t size, void *buf) except -1:
+    (<object>f.fileobj).seek(addr)
+    cdef b = (<object>f.fileobj).read(size)
+    if len(b) != size:
+        return 1
+    cdef unsigned char[:] mview = bytearray(b)
```
[the remark this answers is not in the document]
Comment (llllllllll) [reply h5py/h5fd.pyx]: another thought, some file-like objects support `readinto` which takes the allocated output and just fills it. This would be a big optimization for file-like objects that support this interface.

File: h5py/h5fd.pyx [hunk 6]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+  PyObject* fileobj
+  haddr_t eoa
+
+
+# A minimal subset of callbacks is implemented. Non-essential
+# parameters (dxpl, type) are ignored.
+
+from cpython cimport Py_INCREF, Py_DECREF
+from libc.stdint cimport *
+from libc.stdio cimport *
```
[the remark this answers is not in the document]
Comment (aragilar) [reply h5py/h5fd.pyx]: I'd suggest import e.g. `free` as `stdlib_free`, as that adds clarity as to where free came from. I'd also be wary of something from libc overriding/shadowing a variable (as there are lots of common names in libc), so I'd avoid using `cimport *`.

File: h5py/_hl/files.py (hunk 3 above)
Comment (aparamon) [reply h5py/_hl/files.py]: Do you mean extending `_drivers` init value rather than using `register_driver` call? Yes, I can certainly make such a mini-refactoring.

File: h5py/_hl/files.py (hunk 4 above)
Comment (aparamon) [reply h5py/_hl/files.py]: Should that be a `ValueError` or `H5pyWarning`?

**Comment by aparamon** [comment]
@llllllllll Thank you for in-depth review! Please note that my Cython driver callbacks use `except` keyword machinery, so original Python should be recorded *somewhere* for later retrieval.
I'm not sure what part of exception trace is more valuable for the end user: the original Python fileobj's or HDF5 one. Unfortunately, I'm currently away from my dev machine so I cannot check -- but what happens currently if you pass a "broken" fileobj that raises upon read or seek?

File: h5py/_hl/files.py [hunk 7]
```diff
@@ -342,15 +343,20 @@ def __init__(self, name, mode=None, driver=None,
             with phil:
                 fid = h5i.get_file_id(name)
         else:
-            name = filename_encode(name)
+            if hasattr(name, 'read') and hasattr(name, 'seek'):
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/_hl/files.py]: I personally don't like `h5py.File` **constructor** possessing such a strong side-effect (`read` call), but if there is a consensus I'll add the check.

[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/_hl/files.py]: `write`, `truncate`, and `flush` -- only occasionally. It should really come quite natural for any h5py user that fileobj *must* implement those methods in order to write data.
I added `read` and `seek` safe-guards to make code more readable and self-documented, not to prevent all misuses (in Python, that's impossible). Technically more checks could be added of course -- but is it worth the complexity?

File: h5py/h5fd.pyx [hunk 8]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+
+
+cdef H5FD_fileobj_t *H5FD_fileobj_open(const char *name, unsigned flags, hid_t fapl, haddr_t maxaddr):
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: Could you please elaborate? I think all we can do on failure here is to return `Null`, and that's what's happening.

File: h5py/h5fd.pyx [hunk 9]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+    cdef unsigned long long int p
+    cdef H5FD_fileobj_t *f
+    if sscanf(name, "%llx", &p):
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: But does it mean I'll have to implement the `fapl_*` callbacks, and additionally a way to pass the pointer value to `H5FD_class_t` from Python code? In my personal feeling that will make the code less understandable and so *more* brittle, -- but should be possible technically.

File: h5py/h5fd.pyx [hunk 10]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eoa(const H5FD_fileobj_t *f, H5FD_mem_t type):
+    return f.eoa
+
+cdef herr_t H5FD_fileobj_set_eoa(H5FD_fileobj_t *f, H5FD_mem_t type, haddr_t addr) except -1:
+    f.eoa = addr
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eof(const H5FD_fileobj_t *f, H5FD_mem_t type) except -1:
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: Somehow, it did pass the checks on the test robots :-)
What is the correct way to define the conditionals?
(Please also note that `H5FD_class_t` and `H5FD_t` structs are seemingly not binary-compatible between 1.8 and 1.10, so probably need to address this as well. Do we care much of 1.8 still?)

File: h5py/h5fd.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: My biggest problem with this code was to support older Cythons:
https://github.com/h5py/h5py/pull/1061#issuecomment-402138536
It should be measured, but I'm not really sure `readinto` would produce *major* speed-ups here. The actual file access is usually the bottleneck, not the `memcpy`.

File: h5py/h5fd.pyx [hunk 11]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+    if len(b) != size:
+        return 1
+    cdef unsigned char[:] mview = bytearray(b)
+    memcpy(buf, &mview[0], size)
+    return 0
+
+cdef herr_t H5FD_fileobj_write(H5FD_fileobj_t *f, H5FD_mem_t type, hid_t dxpl, haddr_t addr, size_t size, void *buf) except -1:
+    (<object>f.fileobj).seek(addr)
+    cdef b = bytearray(<unsigned char[:size]>buf)
+    (<object>f.fileobj).write(b)
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: Could you please check it works fine with older Pythons and Cythons?

File: h5py/h5fd.pyx [hunk 12]
```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
...
+info.close = <herr_t (*)(H5FD_t *)>H5FD_fileobj_close
+info.get_eoa = <haddr_t (*)(const H5FD_t *, H5FD_mem_t)>H5FD_fileobj_get_eoa
+info.set_eoa = <herr_t (*)(H5FD_t *, H5FD_mem_t, haddr_t)>H5FD_fileobj_set_eoa
+info.get_eof = <haddr_t (*)(const H5FD_t *, H5FD_mem_t)>H5FD_fileobj_get_eof
+info.read = <herr_t (*)(H5FD_t *, H5FD_mem_t, hid_t, haddr_t, size_t, void *)>H5FD_fileobj_read
+info.write = <herr_t (*)(H5FD_t *, H5FD_mem_t, hid_t, haddr_t, size_t, const void *)>H5FD_fileobj_write
+info.truncate = <herr_t (*)(H5FD_t *, hid_t, hbool_t)>H5FD_fileobj_truncate
+info.flush = <herr_t (*)(H5FD_t *, hid_t, hbool_t)>H5FD_fileobj_flush
+
+fileobj_driver = H5FDregister(&info)
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: Could you please elaborate?

File: h5py/h5p.pyx [hunk 13]
```diff
@@ -1032,6 +1032,16 @@ cdef class PropFAID(PropInstanceID):
         H5Pset_fapl_stdio(self.id)
 
 
+    @with_phil
+    def set_driver(self, hid_t driver_id):
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5p.pyx]: Maybe, but `get_driver` (see below) didn't...

File: h5py/tests/hl/test_file.py [hunk 14]
```diff
@@ -127,3 +131,58 @@ def test_w0(self):
         f = h5py.File(fname, 'w', rdcc_w0=0.25)
         self.assertEqual(list(f.id.get_access_plist().get_cache()),
                          [0, 521, 1048576, 0.25])
+
+
+class TestFileObj(TestCase):
+    def test_BytesIO(self):
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/tests/hl/test_file.py]: I believe the tastes may differ here :-)
But your method will definitely become preferable when we start adding tests for more fileobjs.

File: h5py/tests/hl/test_file.py [hunk 15]
```diff
@@ -127,3 +131,58 @@ def test_w0(self):
         f = h5py.File(fname, 'w', rdcc_w0=0.25)
         self.assertEqual(list(f.id.get_access_plist().get_cache()),
                          [0, 521, 1048576, 0.25])
+
+
+class TestFileObj(TestCase):
```
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/tests/hl/test_file.py]: I will add more tests.

File: h5py/h5fd.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (llllllllll) [reply h5py/h5fd.pyx]: why do we need `bytearray` at all, shouldn't a `memoryview` work fine?

File: h5py/h5fd.pyx (hunk 9 above)
[the remark this answers is not in the document]
Comment (llllllllll) [reply h5py/h5fd.pyx]: right now users can do:
```python
h5py.File('0', driver='fileobj')
```
which will just segfault because it will parse as a pointer and then dereference null. Also, it is not obvious that the object will be alive between the call to `id` and the call to `open`. This happens to be the case now because the driver open call happens in `File.__init__`, and the object must be alive on the stack of the calling frame. Imagine if someone deferred the `open` call to the first use, now nothing has kept the file-like object alive, and `id` may or may not have been reclaimed. This will cause weird bugs later on. By using actual driver arguments we can remove this "serialization" step which makes the required lifetime of the object hard to follow.

Also, on 32 bit systems the pointers will not be long long ints, which means this is a potentially truncating cast.

File: h5py/h5fd.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (llllllllll) [reply h5py/h5fd.pyx]: This copy is happening _after_ any filters have been undone, which means it could be a large decompressed memcpy. `memoryview` (called `buffer` in py2) has been around since before 2.5, it's not a Cython feature. I was thinking just directly doing:
```C
python_view = PyMemoryView_FromMemory(buf, size, PyBUF_WRITE)
fileobj.readinto(python_view)
```

File: h5py/h5fd.pyx (hunk 11 above)
[the remark this answers is not in the document]
Comment (llllllllll) [reply h5py/h5fd.pyx]: Same comment from read: memoryview is just a Python object and this is part of the `file` interface. This is supported for `file` in both Python 3 and 2.

File: h5py/_hl/files.py (hunk 3 above)
Comment (llllllllll) [reply h5py/_hl/files.py]: Yeah, might as well keep all the builtins in one place.

File: h5py/h5fd.pyx (hunk 10 above)
Comment (llllllllll) [inline h5py/h5fd.pyx]: `haddr_t` is unsigned and `get_eof` cannot fail in the hdf5 API. To make sure it cannot fail, other drivers just cache the eof. Currently `-1` gets turned into a vary large but otherwise valid value for the eof.

**Comment by llllllllll** [comment]
Currently, if an object doesn't support `truncate`, Python will segfault on exit if you haven't closed the the the h5py file.

File: h5py/h5fd.pyx (hunk 8 above)
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: Good suggestion, will do.

**Comment by aparamon** [comment]
It was suggested that some performance optimizations are introduced, including
- removing unneeded `memcpy`s,
- specifying non-default `fl_map`,
- using `readinto` if available.

Is it fine if I open a separate PR once the basic implementation is merged, -- or it's preferred that I continue the work in this PR?

File: h5py/h5fd.pyx (hunk 10 above)
Comment (aparamon) [reply h5py/h5fd.pyx]: No, `get_eof` can actually return `HADDR_UNDEF` to signal failure:
https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5FD.c#1226

Comment (llllllllll) [reply h5py/h5fd.pyx]: Interesting, I haven't seen a driver that does this. I stand corrected. I looked into this because when `tell` fails this raises a system error. I wonder if we are not propagating this error through correctly in the Cython wrappers. Could we use the symbolic constant `HADDR_UNDEF` to make this more clear?

**Comment by tacaswell** [comment]
@aparamon I lean towards doing it all in this PR.  It helps keep the discussion focused.

File: h5py/h5fd.pyx (hunk 9 above)
[the remark this answers is not in the document]
Comment (llllllllll) [reply h5py/h5fd.pyx]: > But what is the workflow implying call ... and what user expects from h5py in this case?

I would expect that code to tell me that `str` is not a file-like object or to get an attribute error or some other Python error. No user ever expects a segfault as the result of writing Python code. A segfault might be okay for very low-level performance sensitive APIs like reading into a custom dataspace where verification of the arguments is expensive relative to the desired performance, but for something like opening a file I would never expect a segfault. Many people who depend on h5py are not software engineers by trade and won't see this as obviously incorrect. These same people will likely not be able to debug the issue 


> it is currently important that open happens shortly after id call

Nothing about this is explicit nor obvious from the code as written. For example, imagine we want to move `open` to the first read for some subclass of `File`, like a `LazyFile` or something. Now, this code will have a bad bug when temporaries are passed to the constructor. I agree that it is easy to ensure that the sequence of events happens correctly: just pass the pointer to the code that needs the pointer.


> A value of type `long long int` is guaranteed to be at least 64-bit wide, so it will be able to contain a pointer on all systems.

I agree that the integer can store the pointer, I am not sure that you can cast the `long long int` back to a 32 bit pointer in a compliant way. The only int types guaranteed to be able to convert to and from pointers are `intptr_t` and `uintptr_t`. The tests are passing but the code may be relying on undefined behavior.

> However please don't expect it to be any more robust or fast in practical cases.

None of this is about speed or performance, it is about maintainability and being explicit with the intent and requirements of code. Passing a pointer from Python to C through a string which is converted to an integer and then cast to a `PyObject*` is non-obvious behavior. There is already an API which allows us to just pass the pointer through to this code without passing it through so many conversions.

Another thing I hadn't thought of until now: in pypy, the `id` might not even be a valid pointer. This would then just cast a random integer to a pointer and go. Idk if pypy is a supported platform, but I see windows pypy wheels for the latest release.

[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: The latter argument (pypy support) is pretty valid. I will refactor the code to pass the PyObject as you suggested.

File: h5py/h5fd.pyx (hunk 10 above)
Comment (aparamon) [reply h5py/h5fd.pyx]: Will do. As for Python error propagation, it seems that `H5Epush2` is the way to go (although it will add a lot of boilerplate compared to elegant Cython's `except`).

Comment (llllllllll) [reply h5py/h5fd.pyx]: The hdf5 codebase uses macros for this, maybe we can use them also.

File: h5py/h5fd.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (aparamon) [reply h5py/h5fd.pyx]: But `PyMemoryView_FromMemory()` is only available in later Pythons, according to
https://docs.python.org/3/c-api/memoryview.html#c.PyMemoryView_FromMemory
It would be awesome if someone could help me with optimized (no-copy) and universal (across all supported Pythons, Cythons) variants of `H5FD_fileobj_read`, `H5FD_fileobj_write`.

**Comment by aparamon** [comment]
I have implemented less fragile method of passing fileobj to Cython, added more tests, and fixed exception handling. Please let me know it's moving in the right direction.

I considered using `H5Epush2` instead of Cython `except`, but the former is not currently exposed and the latter is used widely, e.g.
https://github.com/h5py/h5py/blob/07f3c362887563c1040b10d2cbd8cca49ceac1b9/h5py/h5g.pyx#L177

The remaining part is optimized (no-copy) read and write, but that's a detail of implementation (albeit important) and should be easy to achieve.

File: h5py/_hl/files.py (hunk 4 above)
Comment (tacaswell) [reply h5py/_hl/files.py]: Lean towards `TypeError` as that is what python gives you for invalid signatures.

**Comment by aparamon** [comment]
Now `H5FD_fileobj_read()` avoids `memcpy` if `readinto()` is available in the file-like object!
I believe that was the final issue. Any obstacle remaining for the merge?

**Comment by tacaswell** [comment]
attn @epourmal Does anyone from hdf5 group have the bandwidth to look at this?

**Comment by tacaswell** [comment]
@aparamon it looks like you rebase, merged the un-rebased branch back in and then pushed?

**Comment by aparamon** [comment]
@tacaswell Uhmmm I'm not sure :-) I wanted to fix a typo in the already-pushed commit message, so I googled and followed a recipe from stackoverflow (I think). The way the resulting commits appeared in this PR suggested that I probably did something sub-optimally, but I decided not to try fix it in order to avoid even stronger breakage :-)

**Comment by tacaswell** [comment]
I think what you want to do to fix this is 

```
git checkout fileobj
git reset --hard 2b6ee1f8520b220019510d2bebbd96898c4450aa
git cherry-pick 31b739c85443a7ea7b00ea81d38297251688f1dc
git push --force
```

These:
 - make sure you are on the right branch
 - reset your local branch back to the commit with the message " Avoid more memcpy's"
 - cherry-picks the "cosmetic fix" commit on top
 - force pushes your local branch to github (which will eliminate the 'avoid move' commit and the merge commit from the public record).  As re-writing history is in general frowned on (for example we can not re-write the history of h5py master), you have to explicitly ask git to do it.

I can also do this for you if you want.

**Comment by tacaswell** [comment]
@llllllllll Were your comments addressed sufficiently?

**Comment by aparamon** [comment]
@tacaswell it would be of extremely kind of you if you could do the fix on your side!

**Comment by tacaswell** [comment]
Great!  I'll clean up those few commits, let CI run and then merge.

I think between this and the VDS work we should do a new release (2.9)

**Comment by tacaswell** [comment]
@aparamon done.  You will find you have different commits on your local branch than on github.  Either use `reset --hard aparamon/fileobj` (assuming the remote pointing at your github account has your user name) or delete your local branch and then create a new one based on the remote branch.

@aragilar my earlier statement of planning to merge was maybe a bit aggressive, I'll merge it tomorrow unless you protest.

**Comment by tacaswell** [comment]
Also, @aparamon if you want github to think you wrote those commits, add the email address you used to your github account (your icon in the corner -> settings -> emails you can have as many as you want attached).

**Comment by aragilar** [comment]
Code-wise this looks fine to me, but there's no user documentation (under `docs/`) on how to use the new `fileobj` driver though, we'll need to add that (though that can be done in a new PR).

**Comment by epourmal** [comment]
All,

I sencerly apologize for my slow response. Unfortunately, The HDF Group developers don’t have any bandwidth at this point. If an issue has been identified in the HDF5 C library, please don’t hesitate to contact help@hdfgroup.org. This is the fastest way to get our attention.

**Comment by aparamon** [comment]
For the docs, I think [BytesIO](https://docs.python.org/3/library/io.html#io.BytesIO) and [TemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile) (or even [SpooledTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile)) are most natural demonstration examples, with more fancy teasers including [S3 driver](https://s3fs.readthedocs.io) and [transparent encryption](http://www.rfk.id.au/software/filelike/api/filelike.wrappers.html#Decrypt) (do you have more vivid imagination?). Somewhat unexpectedly, the feature can even be used to work-around https://github.com/h5py/h5py/issues/839 !

**Comment by tacaswell** [comment]
Merging as is, docs can be follow up.

Thanks @aparamon !

**Comment by leofang** [comment]
Thanks a lot for the hard work, @aparamon !
